framework/data/BaseDataProvider.php

```diff
@@ -44,6 +50,20 @@
     private $_keys;
     private $_models;
     private $_totalCount;
+    
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if (!isset($this->id)) {
+            if (self::$counter > 0) {
```

Comment:
This type of logic is repeated throughout the framework. It also occurs in Widget. What about moving this logic to a separate Class responsible for counting for reasons of consistency, DRY and less tests.